chainerx_cc/chainerx/backward.cc

```diff
@@ -395,8 +397,10 @@ class BackwardImpl {
                 output_array_nodes_.end()) {
                 if (output_array_node != nullptr) {
                     std::shared_ptr<ArrayBody> body = output_array_node->weak_body().lock();
-                    if (body != nullptr && !body->IsGradRequired(backprop_id_)) {
+                    if (body != nullptr && !retain_grad_ && !body->IsGradRequired(backprop_id_)) {
```

Comment:
* `body->IsGradRequired()` can be called twice.
* In the second evaluation, `body` might be `nullptr`. (test condition is insufficient?)

How about below?:
```c++
if (body != nullptr && !body->IsGradRequired(backprop_id_)) {
    if (retain_grad_) {
        internal::ArrayBody::RequireGrad(body, backprop_id_);
    } else {
        body->ClearGrad(backprop_id_);
    }
    
}
```